sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs

```diff
@@ -266,141 +264,161 @@ public virtual async Task<Response<UploadLogsResult>> UploadAsync<T>(string rule
                         {
-                            Task<BatchUpload> runningTask = runningTasks[i];
+                            Task<Response> runningTask = runningTasks[i].Item1;
                             if (!runningTask.IsCompleted)
                             {
                                 continue;
                             }
-
-                            // Before removing runningTask - check if it has any errors and add to error list
-                            if (runningTask.Result.Error != null)
-                                errors.Add(runningTask.Result.Error);
-
+                            // If current task has an exception, log the exception and add number of logs in this task to failed logs
+                            if (runningTask.Exception != null)
+                            {
+                                exceptions = AddException(
+                                    exceptions,
+                                    runningTask.Exception,
+                                    isRequestFailed: false);
+                                logsFailed += runningTasks[i].Item2;
+                            }
+                            // If current task returned a response that was not a success, log the exception and add number of logs in this task to failed logs
+                            if (runningTask.Result.Status != 204)
+                            {
+                                exceptions = AddException(
+                                    exceptions,
+                                    runningTask.Exception,
+                                    isRequestFailed: true,
+                                    runningTask.Result);
+                                logsFailed += runningTasks[i].Item2;
+                            }
                             // Remove completed task from task list
                             runningTasks.RemoveAt(i);
                             i--;
                         }
                     }
                 }
+                catch (Exception ex)
+                {
+                    // If we have an error, add error from response into exceptions list which represents the errors from all the batches
+                    exceptions = AddException(
+                                    exceptions,
+                                    ex,
+                                    isRequestFailed: false);
+                }
+            }
 
+            try
+            {
                 // Wait for all the remaining blocks to finish uploading
-                await Task.WhenAll(runningTasks).ConfigureAwait(false);
+                await Task.WhenAll(runningTasks.Select(_ => _.Item1).ToList()).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                // If we have an error, add error from response into exceptions list which represents the errors from all the batches
+                exceptions = AddException(
```

Comment:
We don't want to add this one since we will end up looping over all tasks later and adding exceptions.